### ISSUE TITLE
Resampler performance on Raspberry Pi 3B

### ISSUE THREAD

**Issue by gavv** [roc-project/roc#138]
Currently, we can't play out of the box on this board.

Our default resampler configuration is the following:

```
window_interp = 512
window_size = 64
frame_size = 256
```

On this configuration, resampler requires too much CPU and we get constant PulseAudio underruns and sometimes hangups (#135).

Everything works fine if we disable resampler.

Everything works fine if we reduce `window_size` from 64 to 32. This parameter affects number of CPU cycles directly.

Performance is better (but still not acceptable) if we reduce `window_interp` from 512 to 128. This parameter affects the size of `sync_table_`. 512 gives us 128KB, and 128 gives us 32KB. Any value below 128 gives the same CPU usage. At the same time, the board has 32KB L1 cache.

So the summary for this board and our current resampler is:

* only lower `window_size` values are acceptable currently
* lowering `window_interp` so that the `sinc_table_` fits into the L1 cache can help a bit

Now, we should either do some optimizations in resampler or at least adjust defaults (globally or specifically for weak CPUs).

**Comment by gavv** [comment]
Some examples.

Works:

```
$ ./roc-recv -vv -s :10001 -r :10002 --no-resampling --rate 44100
```

Also works:

```
$ ./roc-recv -vv -s :10001 -r :10002 --resampler-window=32
```

Sometimes works, sometimes causes underruns:

```
$ ./roc-recv -vv -s :10001 -r :10002 --resampler-interp=128
```

**Comment by gavv** [comment]
Well, I was able to apply some quick-and-dirty optimizations to `sinc_()` and `resample_()` that reduced average CPU usage to about 60% on this board, so we can play some sound now. Though underruns still happen from time to time (say, once per 2-5 minutes), causing SoX hangs (see #135).

What I did:

* removed all `roc_panic_if()` checks (this helped)
* reduced number of conditional statements and extracted some repeating expressions to variables (didn't check if this step helped)
* reworked `sinc_table_` index calculations so that we always know the index of the next iteration and added cache prefetching for the next index (this helped a lot)
* got a toolchain that supports armv8 target and added `-march=armv8-a` compiler option (this step is required to get prefetching actually working)

I used Linaro toolchain that we're [already using](https://github.com/roc-project/roc/wiki/Building-%28cross-compile%29#raspberry-pi-2-banana-pi-orange-pi-armv7) for other boards. Note that default Raspberry Pi toolchain that supports all Raspberry boards doesn't support armv8.

**Comment by gavv** [comment]
Here is the patch: https://gist.github.com/gavv/10d7f9c4c39e949ef3233c715bc61473

**Comment by gavv** [comment]
Will be fixed by #150.

**Comment by gavv** [comment]
I've prepared a script to play with the memory access pattern when resampler uses sinc table.

Here is the script: https://gist.github.com/gavv/5e28b14b80fa00819258cefb28b5c9c6

And example output:

![115](https://user-images.githubusercontent.com/8436629/53004555-7bbd4c00-3442-11e9-9d42-f5f9d2cf311f.png)

I hoped to improve cache locality by dividing the sinc table into several interleaving probably partially overlapping tables.

I had no success. It seems that the memory access pattern is to irregular and the table step is constantly changing, so this method is not relevant.

**Comment by gavv** [comment]
Here is the summary after playing with resampler in this branch: https://github.com/roc-project/roc/commits/resampler_optimizations2

* removing panics and other conditionals does not help a lot; branch predictor works pretty good on armv8;
* improving cache locality by reverting resampler cycles does not help a lot too; memory access slows down us a bit, but it's not a bottleneck.

However, the original problem was solved in two steps:

* I've enabled -O3 instead of -O2. This gave us 1.5 sessions instead of 0.75 session.
* I've added resampler profiles and enabled "medium" by default. Switching to medium gave us 3.5 sessions instead of 1.5 session.

So now the default setup can play 2-3 sessions on Raspberry Pi 3B. I'm closing this issue for now.

Later we can think about using simd extensions (using gcc/clang builtins or assembly or better Orc) to make resampler less CPU intensive.
